Engine/insert.c

```diff
@@ -1634,28 +1630,63 @@ int useropcd(CSOUND *csound, UOPCODE *p)
     return OK;
 }
 
-/* IV - Sep 1 2002: new opcodes: xin, xout */
-
 int xinset(CSOUND *csound, XIN *p)
 {
   OPCOD_IOBUFS  *buf;
   OPCODINFO   *inm;
   MYFLT **bufs, **tmp;
-  int i;
+  int i, k = 0;
   CS_VARIABLE* current;
+  UOPCODE  *udo;
+  MYFLT parent_sr;
 
   (void) csound;
   buf = (OPCOD_IOBUFS*) p->h.insdshead->opcod_iobufs;
+  buf->iflag = 1;
+  parent_sr = buf->parent_ip->esr;
   inm = buf->opcode_info;
-  bufs = ((UOPCODE*) buf->uopcode_struct)->ar + inm->outchns;
+  udo = (UOPCODE*) buf->uopcode_struct;
+  bufs = udo->ar + inm->outchns;
   tmp = buf->iobufp_ptrs; // this is used to record the UDO's internal vars
   // for copying at perf-time
   current = inm->in_arg_pool->head;
   for (i = 0; i < inm->inchns; i++) {
     void* in = (void*)bufs[i];
     void* out = (void*)p->args[i];
     tmp[i + inm->outchns] = out;
-    current->varType->copyValue(csound, current->varType, out, in);
+    // DO NOT COPY K or A vars
```

Comment:
I think doing type-specific code here is going to cause limitations for supporting all data types, and that support should be done before this is merged into develop. 

What about adding an oversampling or converter argument to copyValue? Seems like there needs to be the input argument var at source SR and local version of var at target SR.  xinset/xoutset/useropcd should target copying the value from source SR to target SR variables. Adding that to the CS_TYPE gets us the ability to handle per data type, and things like UDTs can walk their members to over/undersample.

Reply:
I think that is a good idea but can be done at a second stage, it is beyond the scope of this work. Note that all the type-specific code was there already, I just moved things around (I did not add anything particular). So it seems to me that you are right but it's a different PR.

I would like to merge this as soon as possible as it is holding up other work that depends on the local SR definitions. I am also not sure about the type system and would need to learn more in order to make such changes.